src/SqlBinding/SqlAsyncCollector.cs

```diff
@@ -90,55 +92,284 @@ public async Task FlushAsync(CancellationToken cancellationToken = default)
+            {
+                return $@"
+                    select 
+                        {ColumnName}
+                    from
+                        INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
+                    inner join
+                        INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE ccu on ccu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME AND ccu.TABLE_NAME = tc.TABLE_NAME
+                    where
+                        tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
+                    and
+                        tc.TABLE_NAME = '{tableName}'
+                    and
+                        tc.TABLE_SCHEMA = {schema}";
+            }
+
+            /// <summary>
+            /// Generates SQL query that can be used to retrieve column names & types of a table
+            /// </summary>
+            public static string GetColumnDefinitionsQuery(string schema, string tableName)
+            {
+                return $@"
+                    select	
+	                    {ColumnName}, DATA_TYPE +
+		                    case 
+			                    when CHARACTER_MAXIMUM_LENGTH = -1 then '(max)'
+			                    when CHARACTER_MAXIMUM_LENGTH <> -1 then '(' + cast(CHARACTER_MAXIMUM_LENGTH as varchar(4)) + ')'
+                                when DATETIME_PRECISION is not null then '(' + cast(DATETIME_PRECISION as varchar(1)) + ')'
+			                    when DATA_TYPE in ('decimal', 'numeric') then '(' + cast(NUMERIC_PRECISION as varchar(9)) + ',' + + cast(NUMERIC_SCALE as varchar(9)) + ')'
+			                    else ''
+		                    end as {ColumnDefinition}	
+                    from 
+	                    INFORMATION_SCHEMA.COLUMNS c
+                    where
+	                    c.TABLE_NAME = '{tableName}'         
+                    and
+                        c.TABLE_SCHEMA = {schema}";
+            }
+
+            /// <summary>
+            /// Generates reusable SQL query that will be part of every upsert command.
+            /// </summary>
+            public static string GetMergeQuery(IList<string> primaryKeys, IEnumerable<string> columnNames, string fullTableName)
+            {
+                // Generate the ON part of the merge query (compares new data against existing data)
+                string primaryKeyMatchingQuery = $"ExistingData.{primaryKeys[0]} = NewData.{primaryKeys[0]}";
+                foreach (var primaryKey in primaryKeys.Skip(1))
+                {
+                    primaryKeyMatchingQuery += $" AND ExistingData.{primaryKey} = NewData.{primaryKey}";
+                }
+
+                // Generate the UPDATE part of the merge query (all columns that should be updated)
+                string columnMatchingQuery = string.Empty;
+                foreach (var column in columnNames)
+                {
+                    columnMatchingQuery += $" ExistingData.{column} = NewData.{column},";
+                }
+                columnMatchingQuery = columnMatchingQuery.TrimEnd(',');
```

Comment:
Consider using `StringBuilder` instead of relying on string concatenation here and above. That would be more efficient  compared to relying on string concatenation.